clients/venice-client/src/main/java/com/linkedin/venice/fastclient/StatsAvroGenericStoreClient.java

```diff
@@ -155,24 +155,16 @@ private <R> CompletableFuture<R> recordRequestMetrics(
       double latency = LatencyUtils.getLatencyInMS(startTimeInNS);
       clientStats.recordRequestKeyCount(numberOfKeys);
 
+      boolean exceptionReceived = false;
       if (throwable != null) {
         /**
-         * If the request (both original and retry) failed due to an exception,
-         * just increment the unhealthy counters as the other counters might not
-         * be useful, especially when the counters are currently common for
-         * healthy/unhealthy requests. No new unhealthy specific error
-         * counters for now apart from the below 2.
+         * If the request (both original and retry) failed due to an exception, its
+         * considered an unhealthy request: so only incrementing a subset of metrics.
          */
-        clientStats.recordUnhealthyRequest();
-        clientStats.recordUnhealthyLatency(latency);
-        if (throwable instanceof VeniceClientException) {
-          throw (VeniceClientException) throwable;
-        } else {
-          throw new VeniceClientException(throwable);
-        }
+        exceptionReceived = true;
       }
 
-      if (latency > TIMEOUT_IN_SECOND * Time.MS_PER_SECOND) {
+      if (exceptionReceived || (latency > TIMEOUT_IN_SECOND * Time.MS_PER_SECOND)) {
```

Comment:
Good fix, I'm surprised that we used to not count requests that throws exceptions as unhealthy requests.

Reply:
We were counting that. But just for unhealthy request count and unhealthy latency. We were not counting it any retry related metrics, etc in such scenarios. So the overview of what was happening in the system felt incomplete, so incrementing more metrics in failure cases as well.